cmd/exec-client/main.go

```diff
@@ -0,0 +1,161 @@
+	id := flag.String("id", "", "The process ID. Required for [stop,status,output].")
+	cmd := flag.String("command", "", "The executable to run. Required for [start].")
+	args := flag.String("args", "", "A comma seperated list of arguments. Optional, used for [start].")
+	cpu := flag.Uint64("cpu", 0, "The CPU time in ms allocated to the process. Optional, used for [start].")
+	mem := flag.Uint64("mem", 0, "The memory quota in bytes allocated to the process. Optional, used for [start].")
+	dio := flag.Uint64("io", 0, "The disk read/write limit in bytes/s allocated to the process. Optional, used for [start].")
+
+	flag.Parse()
+	log.Println("Starting exec-client.")
+	validateFlags(*call)
+
+	log.Printf("config: cert=%s", *cert)
+	log.Printf("config: key=%s", *key)
+	log.Printf("config: cas=%s", *cas)
+	log.Printf("config: address=%s", *address)
+
+	cred, err := creds(*cert, *key, *cas)
+	if err != nil {
+		log.Fatalf("Unable to create tls credentials: %v", err)
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	conn, err := grpc.DialContext(ctx, *address, grpc.WithTransportCredentials(cred), grpc.WithBlock())
+	if err != nil {
+		log.Fatalf("Unable to connect to server: %v", err)
+	}
+	defer conn.Close()
+
+	client := pb.NewExecutorClient(conn)
+
+	switch *call {
+	case "start":
+		resp, err := client.Start(ctx, startParams(*cmd, *args, *cpu, *mem, *dio))
+		if err != nil {
+			log.Fatalf("Start call failed: %v", err)
+		}
+		log.Printf("Start call succeeded: %v", resp)
+	case "stop":
+		resp, err := client.Stop(ctx, &pb.ProcessID{Id: *id})
+		if err != nil {
+			log.Fatalf("Stop call failed: %v", err)
+		}
+		log.Printf("Stop call succeeded: %v", resp)
+	case "status":
+		resp, err := client.Status(ctx, &pb.ProcessID{Id: *id})
+		if err != nil {
+			log.Fatalf("Status call failed: %v", err)
+		}
+		log.Printf("Status call succeeded: %v", resp)
+	case "output":
+		stream, err := client.Output(ctx, &pb.ProcessID{Id: *id})
+		if err != nil {
+			log.Fatalf("Output call failed: %v", err)
+		}
+		for {
+			msg, err := stream.Recv()
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				log.Fatalf("Output encountered an error when recieving stream data: %v", err)
+			}
+			log.Printf("stdout: %s\nstderr: %s", msg.Stdout, msg.Stderr)
+		}
+	}
+}
+
+func creds(certFile, keyFile, caFile string) (credentials.TransportCredentials, error) {
+	cert, err := tls.LoadX509KeyPair(certFile, keyFile)
+	if err != nil {
+		return nil, err
+	}
+	p, err := os.ReadFile(caFile)
+	if err != nil {
+		return nil, err
+	}
+	pool := x509.NewCertPool()
+	if ok := pool.AppendCertsFromPEM(p); !ok {
+		return nil, fmt.Errorf("unable to add server ca to pool")
+	}
+	return credentials.NewTLS(&tls.Config{
+		Certificates: []tls.Certificate{cert},
+		RootCAs:      pool,
+		MinVersion:   tls.VersionTLS13,
+	}), nil
+}
+
+func passedFlag(name string) bool {
```

Comment:
For all the validation below, you not only care that the flag was passed, but also that the value was not empty.
For example, the current validation succeeds for `exec-client --call=start --command=""`.

You should check that the values are non-empty below instead.

Reply:
Sure I can add that here, my original plan was just to do this on the server in case another client implementation is used (server would return an error if an empty string is sent as the command)

Reply:
This client validation can be simplified though
You can replace e.g. `if !passedFlag("command")` with `if command == ""` and remove `func passedFlag` entirely. Plus, it's more robust since the compiler will complain when `command` is typo-d, but won't complain if `"command"` is.